tests/shared/src/test/scala/scala/meta/tests/prettyprinters/SyntacticSuite.scala

```diff
@@ -1933,4 +1933,167 @@ class SyntacticSuite extends scala.meta.tests.parsers.ParseSuite {
+      |  }
+      |}
+      """.trim.stripMargin.replace("\n", EOL)
+    )
+    assertEquals(
+      blockStat("{{a}; val b = 1; val c = 2; {d}}").syntax,
+      s"""
+      |{
+      |  {
+      |    a
+      |  }
+      |  val b = 1
+      |  val c = 2
+      |
+      |  {
+      |    d
+      |  }
+      |}
+      """.trim.stripMargin.replace("\n", EOL)
+    )
+
+    def testBlock(t: String, expectedStr: String, needSep: Boolean): Unit = {
+      def expected(sep: String): String =
+        s"""
+        |{
+        |${expectedStr.split("\n").map("  " + _).mkString("", "\n", sep)}
+        |  {
+        |    a
+        |  }
+        |}
+        """.trim.stripMargin.replace("\n", EOL)
+
+      val code = s"{$t;{a}}"
+      val newBlock = templStat(code)
+      val syntax = expected(sep = if (needSep) "\n" else "")
+      // TODO runTestAssert[Stat](code, Some(syntax))(newBlock)
+      assertEquals(
+        newBlock.syntax,
+        syntax
+      )
+      assertEquals(
+        newBlock.structure,
+        templStat(newBlock.syntax).structure
+      )
+      // If it compiles it cannot produce the same tree, otherwise the separator has unnecessary
+      if (needSep)
+        scala.util.Try(templStat(expected(sep = ""))).toOption.foreach { statNoSep =>
+          assertNotEquals(
+            newBlock.structure,
+            statNoSep.structure
+          )
+        }
+    }
+
+    def sep(t: String, expected: String): Unit = testBlock(t, expected, needSep = true)
+    def noSep(t: String, expected: String): Unit = testBlock(t, expected, needSep = false)
+
+    def declTest(f: String => Unit): Unit = Seq("val", "var", "def").foreach(f)
+    def classTest(f: String => Unit): Unit = Seq("class", "object", "trait").foreach(f)
+
+    declTest(k => sep(s"$k foo: Int", s"$k foo: Int"))
+    noSep("type foo", "type foo")
+    declTest(k => sep(s"$k foo: Int = 1", s"$k foo: Int = 1"))
+    declTest(k => noSep(s"$k foo: Int = {1}", s"$k foo: Int = {\n  1\n}"))
+    sep("def a = macro someMacro", "def a = macro someMacro")
+    sep("type foo = Int", "type foo = Int")
+    classTest(k => sep(s"$k Foo", s"$k Foo"))
+    classTest(k => noSep(s"$k Foo {val foo = 1}", s"$k Foo { val foo = 1 }"))
+    classTest(k => sep(s"$k Foo extends Bar", s"$k Foo extends Bar"))
+    classTest(k =>
+      sep(
+        s"$k Foo extends {val foo = 1} with Bar",
+        s"$k Foo extends { val foo = 1 } with Bar"
+      )
+    )
+    classTest(k =>
+      noSep(
+        s"$k Foo extends {val foo = 1} with Bar {val bar = 2}",
+        s"$k Foo extends { val foo = 1 } with Bar { val bar = 2 }"
+      )
+    )
+    sep("this", "this")
+    sep("Foo", "Foo")
+    sep("Foo.bar", "Foo.bar")
+    sep("Foo.bar", "Foo.bar")
+    sep("-10", "-10")
+    noSep("-{10}", "-{\n  10\n}")
+    sep("foo(bar)", "foo(bar)")
+    sep("foo[Bar]", "foo[Bar]")
+    // sep("foo + ()", "foo + ()") // TODO bug #3138 ?
+    sep("foo + bar", "foo + bar")
+    noSep("foo + {bar}", "foo + {\n  bar\n}")
```

Comment:
about the `ApplyInfix` change checking for `Block`: I think it was this one. at least, one of the tests failed for me. not sure how my `testBlock` is different from yours (they look similar) :)

Reply:
Now after the rebase it's failing for me too, it's probably a regression from #3141, do you want me to create a new issue for this?
The case is something like this:
```scala
  test("#3138: infix with block with another following block") {
    val codeOnNextLine =
      """|{
         |  Foo bar { a }
         |  { b }
         |}
         |""".stripMargin
    val syntaxOnNextLine =
      """|{
         |  Foo bar {
         |    a
         |  }
         |  {
         |    b
         |  }
         |}
         |""".stripMargin
    val treeOnNextLine = Term.Block(
      Term.ApplyInfix(
        Term.Name("Foo"),
        Term.Name("bar"),
        Nil,
        List(Term.Block(List(Term.Name("a"))))
      ) :: Term.Block(List(Term.Name("b"))) :: Nil
    )
    runTestAssert[Stat](codeOnNextLine, Some(syntaxOnNextLine))(treeOnNextLine)
  }
```

Right now in the main branch is returning:
```
{
  Foo bar ({
    a
  }) {
    b
  }
}
```

Reply:
> Now after the rebase it's failing for me too, it's probably a regression from #3141, do you want me to create a new issue for this? In v4.7.7 it was parsing this correctly.

Why is it a regression? Why can you do
```
foo bar (1)
{ 2 }
```
but not
```
foo bar { 1 }
{ 2 }
```
or
```
foo bar { (1, 2) }
{ 3 }
```

in fact, looks like both
```
foo bar (1)
{ 2 }
foo bar { 1 }
{ 2 }
```
should be allowed in scala2 but not scala3.

see https://dotty.epfl.ch/docs/internals/syntax.html
and https://www.scala-lang.org/files/archive/spec/2.11/13-syntax-summary.html

Reply:
I think I know the anwser:
If you go back to the initial issue I opened: #3138, you can see that:
```scala
foo bar (1)
{ 2 }
```
is **not** the same as `foo bar (1, 2)` but rather `foo bar ((1){2})` (just one param: `1.apply(2)`)

---

This expression:
```scala
foo bar { 1 }
{ 2 }
```
is actually a block with two terms, `foo bar { 1 }` and `{ 2 }`, no ambiguity.
And you can't place them in the same line (without a `;`)

---

If you try something like `{ (1, 2) } { 3 }` is actually invalid syntax (you would have to place a `;` between), but
```scala
{ (1, 2) }
{ 3 }
```
Are just two blocks in a row (probably inside another block)

---

I think this is just an issue with the scalameta parser, what is your opinion?

Reply:
@RCMartins thank you for that but you are not explaining _WHY_ you think that way.

I ran a little experiment here:
https://scastie.scala-lang.org/HkPoBGmNSwGxSz81IeNBOw

```
def foo1(a: Int)(b: Int) = a + b

def foo2(a: Int) = a + 1

// no errors
0 + foo1 { 1 }
{ 2 }

// error: missing argument list for method foo1
0 + foo1 { 1 }

{ 2 }

// error: Int doesn't take parameters
0 + foo2 { 1 }
{ 2 }
```

which means that `{ 2 }` on a separate line, after `{ 1 }`, was not actual treated as a separate expression.

Reply:
# Part 0 - my useless previous "example"

> If you try something like `{ (1, 2) } { 3 }` is actually invalid syntax (you would have to place a `;` between), but
> ```scala
> { (1, 2) }
> { 3 }
> ```
> Are just two blocks in a row (probably inside another block)

I can see now that this previous example didn't really matter and just added more confusion, so just completely ignore it please :)

# Part 1 - It's just Term.Apply

In my original [issue](https://github.com/scalameta/scalameta/issues/3138) the incorrect expression is a `Term.ApplyInfix`, and I think the parsing problem is only with that particular Term.

I think your experiment tested a different tree that has no issue now or before:
```scala
def foo1(a: Int)(b: Int) = a + b
```

In your example:
```scala
0 + foo1 { 1 }
{ 2 }
```
Is the same as
```scala
0 + ( ( foo1 { 1 } ) { 2 } )
```

So, the "potential problemetic code" is just two `Term.Apply`:
`Term.Apply(Term.Apply(Term.Name("foo1"), List(Term.Block(List(Lit.Int(1))))), List(Term.Block(List(Lit.Int(2)))))`

And with that I have no issue with.

# Part 2 - My example and Infix parsing logic

My example:
```scala
foo bar { 1 }
{ 2 }
```
Seems similar in quick glance to to your example but they are not:
```scala
0 + foo1 { 1 }
{ 2 }
```

With an infix op the right-hand side will only accept one of two things:
1. One Term, that can have braces around (in that case the term is a Term.Block, and any code after `}` should not be part of the ApplyInfix term)
2. Multiple Terms, with a `()` around, and `,` between

In my example, `bar` is the infix op, it will only accept one of two things, in my case its case 1. `{ 1 }` is the term for the right-hand side of the operation. After the right-hand side there is a new line a new `Stat` begins: `{ 2 }`.

In your example, `+` is the infix op, it will only accept one of two things, in your case its case 1. `foo1 { 1 }\n{ 2 }` is the term for right-hand side of the operation. After the right-hand side there's no more code.
*WHY* it parses that way? it first parses `Term.Name("foo1")` then it checks that it has an `{` and creates an `Term.Apply` and then it checks that it has another `{` and creates another `Term.Apply`.

So there is no issue in your example above, because it's just a diferent tree.

# Part 3 - Generating a new example that causes the issue

I can create generate the issue again from your example if I add some braces (and an aditional block `{ 3 }` to trigger the issue):

```scala
0 + {
  foo1 { 1 }
  { 2 }
}
{ 3 }
```

This expression is completely fine in scala, and parses correctly in scalameta 4.7.7, but parses the wrong tree with the latest scalameta version. You can try this new expression in: https://scastie.scala-lang.org/yY7TuHrEQC6K23I5Hj2Btw

As you can see, it's just two Stats, the `+` operator picks up the term on the right side which is: `{ foo1 { 1 } { 2 } }` and finishes parsing that term, it should **NOT** continue into the next line (`{ 3 }`).

# Conclusion

For completeness sake I'll add this two examples in my tests:
```scala
testBlockAddNL("foo {bar}", "foo {\n  bar\n}")
testBlockAddNL("foo\n{bar}\n{baz}", "foo {\n  bar\n} {\n  baz\n}")
```

I hope this helps clarifing the issue, it's not trivial to explain, but I tried my best!
if you need any help generating more problematic examples, let me know, I have less time during the week, but I'll try :)

Reply:
does #3148 address this?

Reply:
@kitbellew yes, I tried merging with the branch locally and it fixes the issue! Thank you :)